text/0000-global-packages.md

```diff
@@ -0,0 +1,151 @@
+
+Within a given Node Major bucket, we expect packages to continue to work without needing to be re-installed, so we can avoid needing to have a separate cache for every individual Node version. This also allows us to reduce the amount of time we spend re-installing packages on Node version switches.
+
+## Interception of Global Installs
+
+When a user runs `npm i -g <package>` or `yarn global add <package>`, instead of throwing a hard error as we currently do, we can make 2 small changes to the workflow but otherwise allow the command to run as expected:
+
+1. Use the `prefix` and `global-folder` options to redirect the install directory to the Volta-managed location. Package managers already provide support for changing the global install directory, so we can leverage that to make sure we always install into the appropriate Node major version cache. We should set these (if possible) using environment variables, so that they can still be overridden by command-line switches from the user in cases where they want to have control over the directories.
+
+2. After the install completes, take inventory of the packages that changed (whether installed, upgraded, or uninstalled) to update an internal manifest of all packages the user has globally installed.
+
+Additionally, if a user calls `volta install <package>`, we can perform the same steps and shell out to `npm i -g <package>` as if the user had typed that directly.
+
+## Automatic Migration When Switching Node Versions
+
+In order to avoid requiring users to keep track of their global packages and manually re-install them whenever they switch Node versions, we can instead manage that for them. Using the manifest generated above whenever a package is updated, we will always know which tools the user has available to them. Then, whenever they switch to a new major version of Node (with `volta install node@<version>`), we can ensure that the full list of packages is made available by running the appropriate `install` commands.
+
+In order to be efficient at this change, we should keep a separate manifest for each major version of Node, so that we can perform a diff with the "primary" manifest and deteremine exactly which changes are needed to get the system in a working state.
+
+This process may take some time, so we should provide some visual feedback to the user to indicate what we're doing (and if possible, progress on how much is left to do). Additionally, we should have a command-line switch to allow users to skip the consolidation if they don't need it or want to make a quick change.
```

Comment:
We should be very cautious around the dev UX of this. I could see an example where a user wants to upgrade their node version and all of a sudden the `volta install node` command takes a lot longer than expected (as it now could be now doing much more than just installing node). Food for though could be either to add some other command to upgrade the globals or to provide a Y/N prompt after the install if the user want to do the migration.

Additionally, we don't want any migration to ever cause an exit code > 1 if a singular migration fails. So we should make sure we make the migration as bullet proof as possible.

Reply:
Another point here is it might be worth talking with some folks about the number of globals they have installed to get some kind of bench mark as to how long a typical migration might take.

We could also store the install times as apart of the manifest file and if the install times cross a certain threshold then we could require the user to agree before hand or alert them somehow what is happening in the background

Reply:
The runtime of installing the separate packages is definitely a concern, and that's why I wanted to call it out here. Specifically, by "visual feedback" in this line, I'm thinking something like a new progress bar with a clear message of what Volta is doing. It definitely shouldn't be in the background or opaque, it should be very clear the user what we are doing at a high level.

I'm also wary of Y/N prompts for a couple of reasons: First, they break scripting without adding a `-y` flag of some king to make them non-interactive. Next, even for situations where the interaction itself is possible, they can really damage UX by being repetitive ("Yes, I _always_ want to update my packages, why do you keep asking?") or by happening after a delay when the user isn't looking at the screen. For that reason, I generally lean much more towards having a reasonable default and then a flag for unusual cases where that doesn't apply. To maintain Volta's features, it makes sense to have re-installing be the default behavior, while opting out is the special-case.

> Additionally, we don't want any migration to ever cause an exit code > 1 if a singular migration fails. So we should make sure we make the migration as bullet proof as possible.

Definitely agreed here! I intended to include something about that, I'll add it now. We should also warn the user if something fails and we expect a tool won't work.